pretyping/coercion.ml

```diff
@@ -441,39 +441,95 @@ let mu env sigma t =
       | None -> sigma, (None, v, IdCoe)
   in aux t
 
+(* Invariant: if [proj] is [Some] then [args_len < npars]
+     (and always [args_len = length rev_args]) *)
+type delayed_app_body = {
+  head : constr;
+  rev_args : constr list;
+  args_len : int;
+  proj : Projection.Repr.t option
+}
```

Comment:
This could be further changed to 
```suggestion
(* Invariant: if [proj] is [Some] then [args_len < npars]
     (and always [args_len = length rev_args]) *)
type delayed_app_body = {
  mutable head : constr;
  mutable rev_args : constr list;
  args_len : int;
  proj : Projection.Repr.t option
}
```
so that repeated forcing is cheap (note that the args_len is the sum of the args in the head and the yet to be applied args)

Reply:
I think we actually need to do that to prevent quadraticity.

Reply:
Let's bench this one too.